pkg/geo/geomfn/rotate_test.go

```diff
@@ -0,0 +1,207 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package geomfn
+
+import (
+	"math"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/stretchr/testify/require"
+	"github.com/twpayne/go-geom"
+)
+
+func TestRotateGeom(t *testing.T) {
+	testCases := []struct {
+		desc     	string
+		input    	geom.T
+		rotRadians	float64
+		expected 	[]float64
+	}{
+		{
+			desc:     "rotate a 2D point where angle is 2Pi",
+			input:    geom.NewPointFlat(geom.XY, []float64{10, 10}),
+			rotRadians:   2*math.Pi,
+			expected: []float64{10, 10},
+		},
+		{
+			desc:     "rotate a 2D point where angle is Pi",
+			input:    geom.NewPointFlat(geom.XY, []float64{10, 10}),
+			rotRadians:   math.Pi,
+			expected: []float64{-10, -10},
+		},
+		{
+			desc:     "rotate a 2D point where angle is Pi/4",
+			input:    geom.NewPointFlat(geom.XY, []float64{10, 10}),
+			rotRadians:   0.785,
+			expected: []float64{0.00563088, 14.1421345},
+		},
+		{
+			desc:     "rotate a line string",
+			input:    geom.NewLineStringFlat(geom.XY, []float64{10, 15, 20, 25}),
+			rotRadians:   0.5,
+			expected: []float64{1.58444, 17.95799, 5.56601, 31.52807},
+		},
+		{
+			desc:     "rotate a line polygon",
+			input:    geom.NewPolygonFlat(geom.XY, []float64{1, 1, 5, 5, 1, 10, 1, 1}, []int{8}),
+			rotRadians:   math.Pi,
+			expected: []float64{-1, -1, -5, -5, -1, -10, -1, -1},
+		},
+		{
+			desc:     "rotate multiple 2D points",
+			input:    geom.NewMultiPointFlat(geom.XY, []float64{10, 10, 20, 20}),
+			rotRadians:   math.Pi/2,
+			expected: []float64{-10, 10, -20, 20},
+		},
+		{
+			desc:     "rotate multiple line strings",
+			input:    geom.NewMultiLineStringFlat(geom.XY, []float64{1, 1, 2, 2, 3, 3, 4, 4}, []int{4, 8}),
+			rotRadians:   (3*math.Pi)/2,
+			expected: []float64{1, -1, 2, -2, 3, -3, 4, -4},
+		},
+		{
+			desc:     "rotate multiple polygon strings",
+			input:    geom.NewMultiPolygonFlat(geom.XY, []float64{1, 1, 2, 2, 3, 3, 1, 1, 4, 4, 5, 5, 6, 6, 4, 4}, [][]int{{8}, {16}}),
+			rotRadians:   math.Pi,
+			expected: []float64{-1, -1, -2, -2, -3, -3, -1, -1, -4, -4, -5, -5, -6, -6, -4, -4},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			actual, err := RotateGeom(tc.input, tc.rotRadians)
+			require.NoError(t, err)
+
+			require.InEpsilonSlice(t, tc.expected, actual.FlatCoords(), 0.00001)
+			require.EqualValues(t, tc.input.SRID(), actual.SRID())
```

Comment:
we should check the type matches as well.

Reply:
fixed in recent commit.

Reply:
not quite -- i should've said matching the shape of the object.
we could have a polygon and linestring with the same flatcoords.

Reply:
fixed in recent commit, I also rebased my fork onto the upstream and squashed my past commits into 1.